File: app/src/main/java/com/ooftf/demo/layout_chain/demo1/CoordinatorLayout.java
package com.ooftf.demo.layout_chain.demo1;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.VelocityTracker;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.OverScroller;
import android.widget.ScrollView;
import android.widget.Scroller;

import androidx.annotation.NonNull;
import androidx.core.view.NestedScrollingParent;
import androidx.core.view.ScrollingView;
import androidx.core.view.ViewCompat;
import androidx.dynamicanimation.animation.FlingAnimation;
import androidx.recyclerview.widget.RecyclerView;

import com.ooftf.basic.utils.DensityUtil;

import java.util.ArrayList;
import java.util.List;

public class CoordinatorLayout extends ViewGroup implements NestedScrollingParent {
    View header = null;
    View body = null;
    View toolbar = null;
    int overLapMax = DensityUtil.INSTANCE.dip2pxInt(54);
    List<ProgressChangeListener> progressChangeListeners = new ArrayList<>();
    public CoordinatorLayout(Context context) {
        super(context);
    }

    public CoordinatorLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CoordinatorLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public CoordinatorLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    {
        addProgressChangedListener(new ProgressChangeListener() {
            @Override
            public void onProgressChange(float progress) {
                layoutBody();
                layoutToolbar();
            }
        });
    }

    public void addProgressChangedListener(ProgressChangeListener listener){
        progressChangeListeners.add(listener);
    }
    public void removeProgressChangedListener(ProgressChangeListener listener){
        progressChangeListeners.remove(listener);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (header == null) {
            header = findViewWithTag("header");
            header.addOnLayoutChangeListener(new OnLayoutChangeListener() {
                @Override
                public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
                    //requestLayout();
                }
            });
        }
        if (body == null) {
            body = findViewWithTag("body");
        }
        if (toolbar == null) {
            toolbar = findViewWithTag("toolbar");
        }
        measureChildWithMargins(header, widthMeasureSpec, 0, heightMeasureSpec, 0);
        measureChildWithMargins(toolbar, widthMeasureSpec, 0, heightMeasureSpec, 0);
        measureChildWithMargins(body, widthMeasureSpec, 0, heightMeasureSpec, toolbar.getMeasuredHeight());
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    int getCanScrollDistance() {
        return header.getMeasuredHeight() - toolbar.getMeasuredHeight();
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int headerLeft = ((MarginLayoutParams) header.getLayoutParams()).leftMargin;
        header.layout(headerLeft, 0, headerLeft + header.getMeasuredWidth(), header.getMeasuredHeight());
        layoutBody();
        layoutToolbar();
    }

    private float getProgress() {
        return getScrollY() / (float) getCanScrollDistance();
    }
    private void layoutBody(){
        int overlap = (int) (overLapMax * (1 - getProgress()));
        int bodyTop = header.getMeasuredHeight() - overlap;
        int bodyLeft = ((MarginLayoutParams) body.getLayoutParams()).leftMargin;
        body.layout(bodyLeft, bodyTop, bodyLeft + body.getMeasuredWidth(), body.getMeasuredHeight() + bodyTop);
    }
    private void layoutToolbar() {
        toolbar.setAlpha(getProgress());
        int toolbarLeft = ((MarginLayoutParams) toolbar.getLayoutParams()).leftMargin;
        toolbar.layout(toolbarLeft, getScrollY(), toolbarLeft + toolbar.getMeasuredWidth(), getScrollY() +toolbar.getMeasuredHeight());
    }

    @Override
    public boolean onStartNestedScroll(View child, View target, int nestedScrollAxes) {
        return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL;
    }

    @Override
    public void onNestedPreScroll(@NonNull View target, int dx, int dy, @NonNull int[] consumed) {
        if (dy > 0) {
            //没有滚动到顶部
            int newCurrentScroll = getScrollY() + dy;
            if (newCurrentScroll < 0) {
                newCurrentScroll = 0;
                consumed[1] = newCurrentScroll - getScrollY();
            } else if (newCurrentScroll > getCanScrollDistance()) {
                newCurrentScroll = getCanScrollDistance();
                consumed[1] = newCurrentScroll - getScrollY();
            } else {
                consumed[1] = dy;
            }
            scrollBy(0, consumed[1]);
        }
    }

    @Override
    public void onNestedScroll(View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed) {

        if (dyUnconsumed < 0) {
            //没有滚动到顶部
            int consumedDy = 0;
            int newCurrentScroll = getScrollY() + dyUnconsumed;
            if (newCurrentScroll < 0) {
                newCurrentScroll = 0;
                consumedDy = newCurrentScroll - getScrollY();
            } else if (newCurrentScroll > getCanScrollDistance()) {
                newCurrentScroll = getCanScrollDistance();
                consumedDy = newCurrentScroll - getScrollY();
            } else {
                consumedDy = dyUnconsumed;
            }
            scrollBy(0, consumedDy);
        }
    }

    OverScroller scroller = new OverScroller(getContext());
    @Override
    public boolean onNestedPreFling(View target, float velocityX, float velocityY) {

        if (velocityY > 0) {
            if(getScrollY()>=getCanScrollDistance()){
                //已经到顶，不消费
                return false;
            }else{
                scroller.fling(0,
                        getScrollY(),
                        (int)velocityX,
                        (int)velocityY,
                        0,
                        0,
                        0,
                        getCanScrollDistance());
                postInvalidate();
                return true;
            }
        }
        return false;
    }
    @Override
    public void computeScroll() {
        //动画未完成
        if (scroller.computeScrollOffset()) {
            scrollTo(0, scroller.getCurrY());
            //触发draw()，必须有这一步
            postInvalidate();
        }
        notifyProgressChange();
    }


    @Override
    public boolean onNestedFling(View target, float velocityX, float velocityY, boolean consumed) {
        int targetOffset = getViewOffset(target);
        if (velocityY < 0&&targetOffset==0) {
            if(getScrollY()>0){
                scroller.fling(0,
                        getScrollY(),
                        (int)velocityX,
                        (int)velocityY,
                        0,
                        0,
                        0,
                        getCanScrollDistance());
                postInvalidate();
                return true;
            }else{
                //已经到顶，不消费
                return false;
            }
        }
        return false;
    }

    int getViewOffset(View target){
        if(target instanceof ScrollingView){
            return  ((ScrollingView) target).computeVerticalScrollOffset();
        }else{
            return target.getScrollY();
        }
    }

    void notifyProgressChange(){
        float progress = getProgress();
        for(ProgressChangeListener listener : progressChangeListeners){
            listener.onProgressChange(progress);
        }
    }
    @Override
    protected ViewGroup.LayoutParams generateLayoutParams(ViewGroup.LayoutParams lp) {
        return new MarginLayoutParams(lp);
    }

    @Override
    protected MarginLayoutParams generateDefaultLayoutParams() {
        return new MarginLayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT);
    }

    @Override
    public LayoutParams generateLayoutParams(AttributeSet attrs) {
        return new MarginLayoutParams(getContext(), attrs);
    }

    static interface ProgressChangeListener{
       void onProgressChange(float progress);
    }
}

File: app/src/main/java/com/ooftf/demo/layout_chain/demo2/CoordinatorLayout.kt
package com.ooftf.demo.layout_chain.demo2

import android.content.Context
import android.util.AttributeSet
import com.ooftf.basic.utils.DensityUtil.dip2pxInt
import android.view.ViewGroup
import androidx.core.view.NestedScrollingParent
import com.ooftf.basic.utils.DensityUtil
import com.ooftf.demo.layout_chain.demo2.CoordinatorLayout.ProgressChangeListener
import android.view.ViewGroup.MarginLayoutParams
import androidx.core.view.ViewCompat
import android.widget.OverScroller
import androidx.core.view.ScrollingView
import android.widget.FrameLayout
import androidx.recyclerview.widget.RecyclerView
import android.util.SparseIntArray
import android.widget.TextView
import android.view.Gravity
import com.ooftf.demo.layout_chain.demo3.PageView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import android.view.MotionEvent
import android.view.View
import com.ooftf.demo.layout_chain.demo4.StickyRecyclerViewLayout
import com.ooftf.demo.layout_chain.demo4.StickyAdapter
import java.util.ArrayList

class CoordinatorLayout : ViewGroup, NestedScrollingParent {
    var header: View? = null
    var body: View? = null
    var toolbar: View? = null
    var overLapMax = dip2pxInt(54f)
    var progressChangeListeners: MutableList<ProgressChangeListener> = ArrayList()

    constructor(context: Context?) : super(context) {}
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
    }

    constructor(
        context: Context?,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes) {
    }

    fun addProgressChangedListener(listener: ProgressChangeListener) {
        progressChangeListeners.add(listener)
    }

    fun removeProgressChangedListener(listener: ProgressChangeListener) {
        progressChangeListeners.remove(listener)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (header == null) {
            header = findViewWithTag("header")
        }
        if (body == null) {
            body = findViewWithTag("body")
        }
        if (toolbar == null) {
            toolbar = findViewWithTag("toolbar")
        }
        measureChildWithMargins(header, widthMeasureSpec, 0, heightMeasureSpec, 0)
        measureChildWithMargins(toolbar, widthMeasureSpec, 0, heightMeasureSpec, 0)
        measureChildWithMargins(
            body,
            widthMeasureSpec,
            0,
            heightMeasureSpec,
            toolbar!!.measuredHeight
        )
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    val canScrollDistance: Int
        get() = header!!.measuredHeight - toolbar!!.measuredHeight

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        layoutHeader()
        layoutBody()
        layoutToolbar()
    }

    private fun layoutHeader() {
        val lp = header!!.layoutParams as MarginLayoutParams
        header!!.layout(
            lp.leftMargin,
            lp.topMargin,
            lp.leftMargin + header!!.measuredWidth,
            lp.topMargin + header!!.measuredHeight
        )
    }

    private val progress: Float
        private get() = scrollY / canScrollDistance.toFloat()

    private fun layoutBody() {
        val lp = body!!.layoutParams as MarginLayoutParams
        val overlap = (overLapMax * (1 - progress)).toInt()
        val bodyTop = header!!.measuredHeight - overlap + lp.topMargin
        body!!.layout(
            lp.leftMargin,
            bodyTop,
            lp.leftMargin + body!!.measuredWidth,
            body!!.measuredHeight + bodyTop
        )
    }

    private fun layoutToolbar() {
        toolbar!!.alpha = progress
        val lp = toolbar!!.layoutParams as MarginLayoutParams
        toolbar!!.layout(
            lp.leftMargin,
            lp.topMargin + scrollY,
            lp.leftMargin + toolbar!!.measuredWidth,
            lp.topMargin + scrollY + toolbar!!.measuredHeight
        )
    }

    override fun onStartNestedScroll(child: View, target: View, nestedScrollAxes: Int): Boolean {
        return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL
    }

    override fun onNestedPreScroll(target: View, dx: Int, dy: Int, consumed: IntArray) {
        if (dy > 0) {
            //没有滚动到顶部
            var newCurrentScroll = scrollY + dy
            if (newCurrentScroll < 0) {
                newCurrentScroll = 0
                consumed[1] = newCurrentScroll - scrollY
            } else if (newCurrentScroll > canScrollDistance) {
                newCurrentScroll = canScrollDistance
                consumed[1] = newCurrentScroll - scrollY
            } else {
                consumed[1] = dy
            }
            scrollBy(0, consumed[1])
        }
    }

    override fun onNestedScroll(
        target: View,
        dxConsumed: Int,
        dyConsumed: Int,
        dxUnconsumed: Int,
        dyUnconsumed: Int
    ) {
        if (dyUnconsumed < 0) {
            //没有滚动到顶部
            var consumedDy = 0
            var newCurrentScroll = scrollY + dyUnconsumed
            if (newCurrentScroll < 0) {
                newCurrentScroll = 0
                consumedDy = newCurrentScroll - scrollY
            } else if (newCurrentScroll > canScrollDistance) {
                newCurrentScroll = canScrollDistance
                consumedDy = newCurrentScroll - scrollY
            } else {
                consumedDy = dyUnconsumed
            }
            scrollBy(0, consumedDy)
        }
    }

    var scroller = OverScroller(context)
    override fun onNestedPreFling(target: View, velocityX: Float, velocityY: Float): Boolean {
        return if (velocityY > 0) {
            if (scrollY >= canScrollDistance) {
                //已经到顶，不消费
                false
            } else {
                scroller.fling(
                    0,
                    scrollY,
                    0,
                    velocityY.toInt(),
                    0,
                    0,
                    0,
                    canScrollDistance
                )
                postInvalidate()
                true
            }
        } else false
    }

    override fun computeScroll() {
        //动画未完成
        if (scroller.computeScrollOffset()) {
            scrollTo(0, scroller.currY)
            //触发draw()，必须有这一步
            postInvalidate()
        }
        notifyProgressChange()
    }

    override fun onNestedFling(
        target: View,
        velocityX: Float,
        velocityY: Float,
        consumed: Boolean
    ): Boolean {
        val targetOffset = getViewOffset(target)
        return if (velocityY < 0 && targetOffset == 0) {
            if (scrollY > 0) {
                scroller.fling(
                    0,
                    scrollY,
                    0,
                    velocityY.toInt(),
                    0,
                    0,
                    0,
                    canScrollDistance
                )
                postInvalidate()
                true
            } else {
                //已经到顶，不消费
                false
            }
        } else false
    }

    fun getViewOffset(target: View): Int {
        return if (target is ScrollingView) {
            (target as ScrollingView).computeVerticalScrollOffset()
        } else {
            target.scrollY
        }
    }

    fun notifyProgressChange() {
        val progress = progress
        for (listener in progressChangeListeners) {
            listener.onProgressChange(progress)
        }
    }

    override fun generateLayoutParams(lp: LayoutParams): LayoutParams {
        return MarginLayoutParams(lp)
    }

    override fun generateDefaultLayoutParams(): MarginLayoutParams {
        return MarginLayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
    }

    override fun generateLayoutParams(attrs: AttributeSet): LayoutParams {
        return MarginLayoutParams(context, attrs)
    }

    interface ProgressChangeListener {
        fun onProgressChange(progress: Float)
    }

    init {
        addProgressChangedListener(object : ProgressChangeListener {
            override fun onProgressChange(progress: Float) {
                layoutBody()
                layoutToolbar()
            }
        })
    }
}